src/main/java/bot/Bot.java

```diff
@@ -0,0 +1,218 @@
+            return;
+        }
+
+        String response = String.format(
+                "Got it. I've added this task:\n  %s\nNow you have %d task(s) in the list.",
+                tasks.get(newTaskIndex).toString(),
+                tasks.size()
+        );
+        ui.printMessage(response);
+    }
+
+    private void handleDeleteTask(String args) throws InvalidTaskIdException {
+        int index = getTaskIndex(args);
+        if (index < 0 || index > tasks.size() - 1) {
+            throw new InvalidTaskIdException(index + 1);
+        }
+        Task deletedTask = tasks.remove(index);
+
+        // Save tasks to disk
+        try {
+            storage.saveTaskList(tasks);
+        } catch (IOException e) {
+            // Revert delete task
+            tasks.add(deletedTask);
+            System.out.println("Task failed to be deleted: " + e.getMessage());
+            return;
+        }
+
+        String response = String.format(
+                "Noted. I've removed this task:\n  %s\nNow you have %d task(s) in the list.",
+                deletedTask.toString(),
+                tasks.size()
+        );
+        ui.printMessage(response);
+    }
+
+    private void handleMarkTask(String args) throws InvalidTaskIdException {
+        int index = getTaskIndex(args);
+        if (index < 0 || index > tasks.size() - 1) {
+            throw new InvalidTaskIdException(index + 1);
+        }
+        boolean isPrevMarked = tasks.isMarked(index);
+        Task markedTask = tasks.mark(index);
+
+        // Save tasks to disk
+        try {
+            storage.saveTaskList(tasks);
+        } catch (IOException e) {
+            // Revert mark task
+            if (isPrevMarked) {
+                tasks.mark(index);
+            } else {
+                tasks.unmark(index);
+            }
+            System.out.println("Task failed to be mark: " + e.getMessage());
+            return;
+        }
+
+        ui.printMessage("Nice! I've marked this task as done:\n" + markedTask);
+    }
+
+    private void handleUnmarkTask(String args) throws InvalidTaskIdException {
+        int index = getTaskIndex(args);
+        if (index < 0 || index > tasks.size() - 1) {
+            throw new InvalidTaskIdException(index + 1);
+        }
+        boolean isPrevMarked = tasks.isMarked(index);
+        Task unmarkedTask = tasks.unmark(index);
+
+        // Save tasks to disk
+        try {
+            storage.saveTaskList(tasks);
+        } catch (IOException e) {
+            // Revert unmark task
+            if (isPrevMarked) {
+                tasks.mark(index);
+            } else {
+                tasks.unmark(index);
+            }
+            System.out.println("Task failed to be unmarked: " + e.getMessage());
+            return;
+        }
+
+        ui.printMessage("OK, I've marked this task as not done yet:\n" + unmarkedTask);
+    }
+
+    private int getTaskIndex(String input) {
```

Comment:
Could be renamed to parseTaskIndexFromInput to indicate that the method parses the task index from user input.